### DISCUSSION TITLE
How to fetch data and update an Alpine object after a custom JavaScript event?

### DISCUSSION THREAD

**Discussion by tumeboy** [alpinejs/alpine#3391] [1. Help] [answered]
I am trying to trigger a custom JavascriptEvent to signal that certain action (a modal is closed) has been done, after which I am trying to fetch data from an API and update an Alpine object. Here's my modal-related code:

```
<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('settings', () => ({

        ...other stuff here...

        postSettings() {
            if (this.settings.userEdited) {
                fetch('/settings', {
                method: 'POST',
                headers: {
                    'X-CSRFToken': document.head.querySelector('meta[name=csrf-token]').content
                },
                body: new FormData(settingsForm)
                })
                .then(response => response.text())
                .then(text => this.settingsId = text)
                .finally(() => this.$dispatch('settingsUpdated'))
            }  
            },
            
        }))
    })
</script>

<section style="width: 100%;" x-data="settings" @keydown.escape="settingsModalOpen = false; postSettings();">

...Rest of the HTML related to the modal...
```
I'm developing a Django web app, and the modal is included to other templates. In another specific template, in which the modal is included I have a script which is listening for event 'settingsUpdated':
```
<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('results', () => ({

            results: [],

            init() {
                this.getResultsValues()
                window.addEventListener('settingsUpdated', this.getResultsValues)
            },

            getResultsValues() {
            fetch("/api/results")
            .then(response => response.json())
            .then(json => this.results = json)
            },

        }))
    })
</script>
```
The issue I'm facing is that the ```this.results``` does not get updated if the event is triggered. I can see from the browser developer tools tab that the event is clearly 'catched' and a GET request is sent to /settings but for some reason the Alpine object ```this.results``` is not updated after the data is fetched. The same data is succesfully fetched and mounted to the object when the page is initialized (```init() {this.getResultsValues()...}```) but for some reason the same function does not update the results object in case the function is triggered by an event (```window.addEventListener('settingsUpdated', this.getResultsValues)```)

P.S. I left the same question also in [StackOverflow](https://stackoverflow.com/q/75249895/8405435) if someone is more comfortable on discussing about the issue there.

**Comment by ekwoka** [comment, marked as answer]
The issue is `this` changing.

```
window.addEventListener('settingsUpdated', this.getResultsValues)
```

should be changes to either

```
window.addEventListener('settingsUpdated', this.getResultsValues.bind(this))
// or
window.addEventListener('settingsUpdated', () => this.getResultsValues())
```

But also, event BETTER is to do
```
@settings-updated.camel.window="getResultsValues"
```

on the element with the `x-data`

additionally, I'd recommend learning async await to keep your syntax cleaner and clearer

**Reply by tumeboy** [reply]
Thank you very much. I tried the two latter suggestions, and both seemed to work. Initially I tried `@settingsUpdated.window="getResultsValues"` but did not figure it out that camelCase doesn't work without additional arguments. Now that I look, it seems to be clearly expressed in the Alpine.js documentation.

**Reply by ekwoka** [reply]
Yup, the docs are good, and it does make sense when you know how HTML works, but it can be easy to just think it would work in camelcase.
